### PR TITLE
Phase 6: Progressive Surface Focus Schedule — Curriculum Learning

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current training applies a high `surf_weight` (dynamically clamped [5, 50]) from the very first epoch. This means the network receives a heavily amplified surface gradient signal before it has learned any useful bulk-flow representation.

**The problem:** Early in training, predictions are random. High surface weight forces the network to immediately optimize surface nodes — but without a good bulk-flow backbone, surface predictions are built on a poor foundation. This is analogous to training a segmentation model with full boundary weight from epoch 1, before the backbone has learned shape features.

**Curriculum learning solution:** Start with `surf_weight=1.0` (equal treatment of surface and volume nodes) and ramp up to the target value over a warm-in period. This forces the model to first build a coherent global flow field, then refine the surface predictions on that solid foundation.

**Distinct from existing mechanisms:**
- `tandem_ramp`: ramps tandem surface loss contribution from 0→1 over epochs 10-50 — this is tandem vs single-foil weighting, NOT surface vs volume
- Dynamic `surf_weight`: already adapts based on loss ratio — the proposal adds a curriculum schedule ON TOP of this
- `vol_ramp_epochs`: ramps volume weight, not surface weight

**Motivation from recent findings:**
- Deep supervision (#2097) showed p_in improves -1.7% when intermediate features get direct surface supervision — suggesting better backbone features help surface predictions
- The progressive schedule achieves a similar effect by letting the backbone develop freely before amplifying surface loss

## Instructions

**Step 1: Add flags** to argparse:
```python
parser.add_argument('--surf_weight_ramp_epochs', type=int, default=0,
                    help='Epochs to ramp surf_weight from 1.0 to dynamic target (0=disabled)')
```
Add to Config dataclass:
```python
surf_weight_ramp_epochs: int = 0   # 0 = disabled
```

**Step 2: In the training loop**, find where `surf_weight` is computed (around the dynamic surf_weight assignment). Add the curriculum schedule:

```python
# Progressive surface focus schedule
if cfg.surf_weight_ramp_epochs > 0 and epoch < cfg.surf_weight_ramp_epochs:
    # dynamic_surf_weight is the normal dynamic value (from loss ratio, clamped [5, 50])
    ramp_frac = epoch / max(cfg.surf_weight_ramp_epochs, 1)
    # Linear interpolation from 1.0 to dynamic target
    surf_weight = 1.0 + (dynamic_surf_weight - 1.0) * ramp_frac
else:
    surf_weight = dynamic_surf_weight  # existing behavior
```

Where `dynamic_surf_weight` is the existing value: `max(5.0, min(50.0, prev_vol_loss / max(prev_surf_loss, 1e-8)))`.

That's it — ~8 lines.

**Step 3: Log the effective surf_weight** for monitoring:
```python
if wandb.run:
    wandb.log({"train/effective_surf_weight": surf_weight}, step=global_step)
```

**Experiment runs** (2 ramp durations × 2 seeds):

```bash
# Ramp over 40 epochs (25% of training), seed 42
python train.py --agent nezuko \
  --wandb_name "nezuko/surframp-40ep-s42" \
  --wandb_group phase6/progressive-surface \
  --surf_weight_ramp_epochs 40 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3

# Ramp over 40 epochs, seed 73
python train.py --agent nezuko \
  --wandb_name "nezuko/surframp-40ep-s73" \
  --wandb_group phase6/progressive-surface \
  --surf_weight_ramp_epochs 40 --seed 73 \
  [same flags]

# Ramp over 80 epochs (50% of training), seed 42
python train.py --agent nezuko \
  --wandb_name "nezuko/surframp-80ep-s42" \
  --wandb_group phase6/progressive-surface \
  --surf_weight_ramp_epochs 80 --seed 42 \
  [same flags]

# Ramp over 80 epochs, seed 73
python train.py --agent nezuko \
  --wandb_name "nezuko/surframp-80ep-s73" \
  --wandb_group phase6/progressive-surface \
  --surf_weight_ramp_epochs 80 --seed 73 \
  [same flags]
```

W&B group: `phase6/progressive-surface`

**What to report:**
- Surface MAE (p_in, p_oodc, p_tan, p_re) per run
- `train/effective_surf_weight` curve — does the ramp look correct?
- Whether 40ep or 80ep ramp is better
- Val/loss convergence profile — does the ramp change convergence speed?

## Baseline

Current best (16-seed ensemble, PR #2093):

| Metric | Baseline |
|--------|----------|
| p_in   | **12.1** |
| p_oodc | **6.6**  |
| p_tan  | **29.1** |
| p_re   | **5.8**  |

Single-model references:
- seed=42: p_in≈12.71, p_oodc≈8.13
- seed=73: p_in≈12.2, p_oodc≈7.59

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent nezuko \
  --wandb_name "nezuko/baseline-s42" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — Progressive Surface Focus

### Setup
- 4 runs: 2 ramp durations (40ep, 80ep) × 2 seeds (42, 73)
- W&B group: `phase6/progressive-surface`
- All runs completed ~157-158 epochs (180-min timeout)

### Individual Run Results (Surface MAE — Pressure)

| Run | W&B ID | Ramp | Seed | p_in | p_oodc | p_tan | p_re |
|-----|--------|------|------|------|--------|-------|------|
| surframp-40ep-s42 | `7n8v8nlh` | 40ep | 42 | 13.334 | 7.927 | 30.574 | 6.487 |
| surframp-40ep-s73 | `ui8t0t7g` | 40ep | 73 | 12.912 | 7.917 | 29.349 | 6.638 |
| surframp-80ep-s42 | `mqu6c2vd` | 80ep | 42 | 13.396 | 7.690 | 29.716 | 6.291 |
| surframp-80ep-s73 | `vkrnehh2` | 80ep | 73 | 12.854 | 8.093 | 29.864 | 6.347 |

### 2-Seed Means vs Baseline (8-seed mean: p_in=13.03, p_oodc=7.83, p_tan=30.29, p_re=6.45)

| Variant | p_in | p_oodc | p_tan | p_re |
|---------|------|--------|-------|------|
| 40ep ramp mean | 13.123 (+0.7%) | 7.922 (+1.2%) | 29.962 (-1.1%) | 6.563 (+1.8%) |
| 80ep ramp mean | 13.125 (+0.7%) | 7.892 (+0.8%) | 29.790 (-1.7%) | 6.319 (-2.0%) |
| Baseline 8-seed mean | 13.03 | 7.83 | 30.29 | 6.45 |

### Analysis

**Mixed results — no clear improvement over baseline.**

**40-epoch ramp:** Slight improvement on p_tan (-1.1%) but regressions on p_in (+0.7%), p_oodc (+1.2%), and p_re (+1.8%). Overall worse than baseline.

**80-epoch ramp:** Better than 40ep variant. Shows improvement on p_tan (-1.7%) and p_re (-2.0%), but slight regression on p_in (+0.7%) and p_oodc (+0.8%). The p_re improvement is noteworthy but comes at the cost of p_in and p_oodc.

**Key observations:**
1. The progressive ramp does help with tandem transfer (p_tan) — both variants improve it. This makes intuitive sense: gradual surface focus lets the model learn volume features first, which benefits transfer.
2. The 80ep ramp is strictly better than 40ep across most metrics, suggesting longer ramps may be beneficial.
3. However, p_in and p_oodc regress in both variants, suggesting the default dynamic surf_weight scheduling from epoch 0 is already well-tuned for in-distribution performance.
4. All results are within normal seed variance (~±0.4 for p_in, ±0.2 for p_oodc), so the differences may not be statistically significant with only 2 seeds.

### Suggested Follow-ups
1. If the advisor wants to explore further: try an even longer ramp (120ep) to see if the trend continues, or try ramping to a fixed surf_weight target instead of the dynamic one.
2. The tandem transfer improvement could compound with other changes — consider combining progressive ramp with a different architecture or loss formulation.
3. Consider closing this direction if the focus should remain on approaches that improve all metrics uniformly.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2110 because ADVISOR: Good execution, nezuko — 4 clean runs, well-structured analysis and insightful observations.

**Decision: Closing as dead end.**

**Why:** The primary metric p_in regresses +0.7% in both ramp variants (40ep and 80ep). While the 80ep ramp shows mild improvement on p_tan (-1.7%) and p_re (-2.0%), these gains are within seed variance (±0.47 for p_tan, ±0.05 for p_re) and don't compensate for the p_in and p_oodc regressions.

**What this tells us:** The existing dynamic surf_weight scheduling from epoch 0 is already well-tuned. The hypothesis that the network needs a 'backbone-first' phase before surface amplification is not supported — the EMA + cosine schedule already provides enough training stability in early epochs. Delaying surface focus gives the model less surface gradient signal overall, which directly hurts p_in.

**Your observation about tandem transfer is interesting** — the gradual ramp helping p_tan makes theoretical sense (better bulk features → better tandem generalization). But we can't trade p_in for p_tan. This insight might be useful if combined with a per-metric loss formulation in a future experiment.

Well done on the thorough analysis. Moving you to a new assignment now..
